src/material/sidenav/drawer.ts

```diff
@@ -339,17 +345,18 @@ export class MatDrawer implements AfterContentInit, AfterContentChecked, OnDestr
   }
 
   /**
-   * If focus is currently inside the drawer, restores it to where it was before the drawer
-   * opened.
+   * If focus is currently inside the drawer, or the drawer is closing via backdrop click,
+   * restore focus to the element that was focused before the drawer opened. This respects
+   * the opened via focus origin that can be optionally specified when opening the drawer.
    */
   private _restoreFocus() {
     if (!this.autoFocus) {
       return;
     }
 
     const activeEl = this._doc && this._doc.activeElement;
-
-    if (activeEl && this._elementRef.nativeElement.contains(activeEl)) {
+    if (this._isClosingViaBackdropClick ||
```

Comment:
This feels a little side-effect-ey. Wouldn't it be better if we moved out the toggling logic into a helper method which has a parameter that determines whether to restore focus. E.g.

```
_closeViaBackdrop() {
  this._toggle(true);
}

toggle() {
  this._toggle(false);
}

_toggle(restoreFocus: boolean) {
   // Closing logic goes here
  if (restoreFocus) {
    this._restoreFocus();
  }
}
```

Reply:
I was slightly concerned about this too, but couldn't think of a case where this would break. I went with the flag as that requires fewest changes and `restoreFocus` seemed already reliant on slide-effects anyway. e.g. `this.autoFocus` being true, reliance on `document.activeElement`.

Maybe this is the chance though to clean this up. I'll experiment with it.

Reply:
Ah. I remember why I didn't do it that way. It was quite similar to #9257. We'd not go through `MatDrawer#toggle´ anymore and that could break tests. I think it would be worth trying against g3 though.